loopd/daemon.go

```diff
@@ -161,75 +179,184 @@ func daemon(config *config, lisCfg *listenerCfg) error {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	return nil
+}
+
+// startEventHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startEventHandlers() {
+	d.mainCtx, d.mainCtxCancel = context.WithCancel(context.Background())
+
+	// Start the swap client itself.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+
+		log.Infof("Starting swap client")
+		err := d.impl.Run(d.mainCtx, d.statusChan)
+		if err != nil {
+			log.Error(err)
 		}
+		log.Infof("Swap client stopped")
 
-		err = grpcServer.Serve(grpcListener)
+		// In case we got here because the swap client crashed, we need
+		// to shutdown as cleanly as possible. If we got here because of
+		// a normal shutdown, stop will prevent itself from running
+		// twice.
+		err = d.Stop()
 		if err != nil {
 			log.Error(err)
 		}
 	}()
 
-	interruptChannel := make(chan os.Signal, 1)
-	signal.Notify(interruptChannel, os.Interrupt)
+	// Start a goroutine that broadcasts swap updates to clients.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
 
-	// Run until the users terminates loopd or an error occurred.
-	select {
-	case <-interruptChannel:
-		log.Infof("Received SIGINT (Ctrl+C).")
+		log.Infof("Waiting for updates")
+		d.processStatusUpdates(d.mainCtx)
+	}()
+}
 
-		// TODO: Remove debug code.
-		// Debug code to dump goroutines on hanging exit.
-		go func() {
-			time.Sleep(5 * time.Second)
-			_ = pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
-		}()
+// Stop tries to gracefully shut down the daemon.
+func (d *Daemon) Stop() error {
+	// Trying to stop the daemon multiple times isn't disallowed, we just
+	// need to make sure the logic itself is only run once.
+	if atomic.AddInt32(&d.stopped, 1) != 1 {
+		return nil
+	}
 
-		cancel()
-	case <-mainCtx.Done():
+	// First of all, cancel the main context that all event handlers are
+	// using. This should stop all swap activity and all event handlers
+	// should exit.
+	if d.mainCtxCancel != nil {
+		d.mainCtxCancel()
 	}
 
-	wg.Wait()
+	// As there is no swap activity anymore, we can shutdown the gRPC and
+	// HTTP servers now.
+	var err error
+	log.Infof("Stopping gRPC server")
+	if d.grpcServer != nil {
+		d.grpcServer.Stop()
+	}
+	log.Infof("Stopping REST server")
+	if d.restServer != nil {
+		// Don't return the error here, we first want to give everything
+		// else a chance to shut down cleanly.
+		err = d.restServer.Close()
+	}
+	if d.restCtxCancel != nil {
+		d.restCtxCancel()
+	}
 
-	return nil
+	// Next, shut down the connections to lnd and the swap server.
+	if d.lnd != nil {
+		d.lnd.Close()
+	}
+	if d.clientCleanup != nil {
+		d.clientCleanup()
+	}
+
+	// The servers should already have stopped gracefully. If not, we
+	// force them by closing its underlying listeners.
```

Comment:
How can this happen?

Reply:
It shouldn't. There was a `defer` before. But looking at `d.grpcServer.Stop()` and `d.restServer.Close()` they both close the underlying listeners in any case.